### PR TITLE
Initial version of auto-detection of terminal width.

### PR DESCRIPTION
**Pull request by dawngerpony**
This appears to be working ok for Python 3.

### CODE REVIEW

**Comment by dawngerpony** [comment]
Update: it seems that the behaviour of `pprint` is different in Python 2.7 compared to 3.9. I can't make `pprint` wrap a long string across multiple lines in 2.7, so it doesn't appear to be respecting the `width` parameter. As such, I can't make the test that I have written pass in both 2.7 and 3.x. Any thoughts?

**Comment by alexmojaki** [comment]
Thanks for contributing!

IMO it's not worth spending significant time on perfecting 2.7 since it's well past EOL. I haven't looked carefully but it sounds like you've made a significant improvement for 3.x and not made anything worse for 2.7 and that's great. You can write something like `expected = '...' if sys.version[0] == 2 else '...'` or just `@unittest.skipIf(sys.version[0] == 2)` on the whole test function. Follow your heart.

**Comment by dawngerpony** [comment]
> Thanks for contributing!

You're very welcome! I love your icecream package and am happy to give back.

> IMO it's not worth spending significant time on perfecting 2.7 since it's well past EOL. I haven't looked carefully but it sounds like you've made a significant improvement for 3.x and not made anything worse for 2.7 and that's great. You can write something like `expected = '...' if sys.version[0] == 2 else '...'` or just `@unittest.skipIf(sys.version[0] == 2)` on the whole test function. Follow your heart.

That's awesome thanks. Unfortunately I had to skip the entire test because there's an assertion inside `parseOutputIntoPairs()` which is tripping up, and I can't skip that assertion just for one test.

But the test is committed, I'll mark this PR as ready.

File: icecream/icecream.py [hunk 1]
```diff
@@ -154,15 +156,40 @@ def format_pair(prefix, arg, value):
...
     s = s.replace('\\n', '\n')  # Preserve string newlines in output.
     return s
 
 
+def columns():
+    """ Returns the number of columns that this terminal can handle. """
+    width = DEFAULT_LINE_WRAP_WIDTH
+    try:
+        # TODO come up with a more elegant solution than subtracting
+        #  a seemingly random number of characters.
```
Comment (dawngerpony) [inline icecream/icecream.py:171]: Any thoughts about this overflow thingy? I just realised that the extra characters might come from the `ic |` prefix.

Comment (alexmojaki) [reply icecream/icecream.py:170]: That's probably what it is, yes. Have this function return the full width, then deal with the length of the prefix inside the class. Remember there's both the `ic |` prefix (which can be configured) as well as the `argPrefix`.

File: icecream/icecream.py [hunk 2]
```diff
@@ -154,15 +156,39 @@ def format_pair(prefix, arg, value):
...
 
-def argumentToString(obj):
-    s = DEFAULT_ARG_TO_STRING_FUNCTION(obj)
+def argumentToString(obj, width=DEFAULT_LINE_WRAP_WIDTH):
+    s = DEFAULT_ARG_TO_STRING_FUNCTION(obj, width=width)
     s = s.replace('\\n', '\n')  # Preserve string newlines in output.
     return s
 
 
+def columns():
```
Comment (alexmojaki) [inline icecream/icecream.py:165]: Please give this a more descriptive name.

File: icecream/icecream.py [hunk 3]
```diff
@@ -154,15 +156,39 @@ def format_pair(prefix, arg, value):
...
+    """ Returns the number of columns that this terminal can handle. """
+    width = DEFAULT_LINE_WRAP_WIDTH
+    try:
+        # TODO come up with a more elegant solution than subtracting
+        #  a seemingly random number of characters.
+        width = os.get_terminal_size().columns - COLUMN_OVERFLOW
+    except OSError:  # Not in TTY
+        pass
+    except AttributeError:  # Python 2.x
```
Comment (alexmojaki) [inline icecream/icecream.py:174]: This looks a bit too broad and prone to hiding real errors. I'd prefer either checking the Python version or checking for the presence of the relevant attribute.

File: icecream/icecream.py [hunk 4]
```diff
@@ -154,15 +156,39 @@ def format_pair(prefix, arg, value):
...
     return s
 
 
+def columns():
+    """ Returns the number of columns that this terminal can handle. """
+    width = DEFAULT_LINE_WRAP_WIDTH
+    try:
+        # TODO come up with a more elegant solution than subtracting
+        #  a seemingly random number of characters.
+        width = os.get_terminal_size().columns - COLUMN_OVERFLOW
```
Comment (alexmojaki) [inline icecream/icecream.py:171]: The docs https://docs.python.org/3/library/os.html#os.get_terminal_size suggest using shutil (https://docs.python.org/3/library/shutil.html#shutil.get_terminal_size) instead, and looking at that function I think that's a good idea.

File: tests/test_icecream.py [hunk 5]
```diff
@@ -518,3 +519,13 @@ def testColoring(self):
...
+
+    @unittest.skipIf(int(sys.version[0]) == 2, "pprint doesn't wrap long strings in Python 2.7")
+    def testLineLengthTen(self):
+        """ Test a specific line wrap width. """
+        ic.lineWrapWidth = 10
+        s = "123456789 1234567890"
+        with disableColoring(), captureStandardStreams() as (out, err):
+            ic(s)
+        pair = parseOutputIntoPairs(out, err, 2)[0]
+        self.assertEqual(pair, [('s', "('123456789 '\n '1234567890')")])
```
Comment (alexmojaki) [inline tests/test_icecream.py:531]: I think using `parseOutputIntoPairs` makes this test worse. I suggest:

1. Checking the whole `err.getvalue()` like in `testMultilineContainerArgs`.
2. Testing the effect of `lineWrapWidth` on big structures like lists and dicts, not just strings. I expect those to still get wrapped in Python 2. `testMultilineContainerArgs` is close to doing this but it's not testing the width configuration.
3. Choosing the expected string based on the version. That should be easy to do and then the test doesn't have to be skipped.

**Review by alexmojaki — CHANGES_REQUESTED** [review]
Overall I like the approach and I appreciate the PR. The `supports_param` idea is clever.

File: icecream/icecream.py (hunk 2 above)
Comment (dawngerpony) [reply icecream/icecream.py]: Done.

File: icecream/icecream.py (hunk 3 above)
Comment (dawngerpony) [reply icecream/icecream.py]: Done.

File: icecream/icecream.py (hunk 4 above)
Comment (dawngerpony) [reply icecream/icecream.py]: Done.

File: tests/test_icecream.py (hunk 5 above)
Comment (dawngerpony) [reply tests/test_icecream.py]: Done. Let me know if you want any more tests (e.g. one with a `dict`).

File: icecream/icecream.py (hunk 1 above)
Comment (dawngerpony) [reply icecream/icecream.py]: Let me know what you think of my new solution.

**Comment by dawngerpony** [comment]
@alexmojaki thanks for the review! I've tried to implement your suggestions, which actually led to a different implementation of `detect_terminal_width` for Python 2.x (using a backport). Let me know what you think 🙏

File: icecream/icecream.py [hunk 6]
```diff
@@ -154,15 +155,38 @@ def format_pair(prefix, arg, value):
...
+    s = DEFAULT_ARG_TO_STRING_FUNCTION(obj, width=width)
     s = s.replace('\\n', '\n')  # Preserve string newlines in output.
     return s
 
 
+def detect_terminal_width(prefix, default=DEFAULT_LINE_WRAP_WIDTH):
+    """ Returns the number of columns that this terminal can handle. """
+    width = default
+    try:
+        if hasattr(shutil, "get_terminal_size"):
```
Comment (alexmojaki) [inline icecream/icecream.py:168]: I'd prefer to have near the top of the file:

```
try:
    from shutil import get_terminal_size
except ImportError:
    try:
        from backports.shutil_get_terminal_size import get_terminal_size
    except ImportError:
        def get_terminal_size():
            # get COLUMNS environment variable
```

icecream should still work if `backports.shutil_get_terminal_size` isn't installed, e.g. if the user chooses to uninstall it even after installing it automatically.

File: icecream/icecream.py [hunk 7]
```diff
@@ -154,15 +155,38 @@ def format_pair(prefix, arg, value):
...
+            width = shutil.get_terminal_size().columns
+        else:  # Python 2.x doesn't support get_terminal_size
+            from backports.shutil_get_terminal_size import get_terminal_size
+            width = get_terminal_size().columns
+    except OSError:  # Not in TTY
```
Comment (alexmojaki) [inline icecream/icecream.py:173]: Looks like the shutil version doesn't raise OSError any more. But maybe the whole thing should be wrapped in `except Exception` just in case something weird happens.

Also you should pass the default width as a fallback to `get_terminal_size`. Otherwise it's going to use its own default fallback (80 columns) when the other methods fail.

File: icecream/icecream.py [hunk 8]
```diff
@@ -173,6 +197,8 @@ def __init__(self, prefix=DEFAULT_PREFIX,
         self.includeContext = includeContext
         self.outputFunction = outputFunction
         self.argToStringFunction = argToStringFunction
+        self.passWidthParam = supports_param(self.argToStringFunction)
+        self.lineWrapWidth = detect_terminal_width(self.prefix, DEFAULT_LINE_WRAP_WIDTH)
```
Comment (alexmojaki) [inline icecream/icecream.py:201]: This still doesn't account for the width of `argPrefix`, or changing the prefix with `configureOutput`.

File: setup.py [hunk 9]
```diff
@@ -103,6 +103,7 @@ def run_tests(self):
         'pygments>=2.2.0',
         'executing>=0.3.1',
         'asttokens>=2.0.1',
+        'backports.shutil-get-terminal-size==1.0.0',
```
Comment (alexmojaki) [inline setup.py:106]: This should only be required in Python 2.

File: tests/test_icecream.py [hunk 10]
```diff
@@ -32,6 +33,10 @@
 c = 3
 
 
+def isPython2():
+    return int(sys.version[0]) == 2
```
Comment (alexmojaki) [inline tests/test_icecream.py:37]:
```suggestion
    return sys.version_info[0] == 2
```

File: tests/test_icecream.py [hunk 11]
```diff
@@ -464,6 +470,7 @@ def testSingleTupleArgument(self):
         self.assertEqual(pair, ('(a, b)', '(1, 2)'))
 
     def testMultilineContainerArgs(self):
+        ic.lineWrapWidth = icecream.DEFAULT_LINE_WRAP_WIDTH
```
Comment (alexmojaki) [inline tests/test_icecream.py:473]: Is this still needed now you have it in setUp?

File: tests/test_icecream.py [hunk 12]
```diff
@@ -518,3 +525,33 @@ def testColoring(self):
             ic({1: 'str'})  # Output should be colored with ANSI control codes.
 
         assert hasAnsiEscapeCodes(err.getvalue())
+
+    def testStringWithLineLengthOfTen(self):
```
Comment (alexmojaki) [inline tests/test_icecream.py:529]: These test names are weird lol

File: tests/test_icecream.py [hunk 13]
```diff
@@ -518,3 +525,33 @@ def testColoring(self):
...
+            ic(s)
+        actual = err.getvalue().strip()
+        if isPython2():
+            self.assertEqual(actual, "ic| s: '123456789 1234567890'")
+        else:
+            expected = textwrap.dedent("""
+            ic| s: ('123456789 '
+                    '1234567890')
+            """).strip()
+            self.assertEqual(actual, expected)
```
Comment (alexmojaki) [inline tests/test_icecream.py:543]:
```suggestion
        if isPython2():
            expected = "ic| s: '123456789 1234567890'"
        else:
            expected = textwrap.dedent("""
            ic| s: ('123456789 '
                    '1234567890')
            """).strip()
        self.assertEqual(actual, expected)
```

File: tests/test_icecream.py (hunk 11 above)
Comment (dawngerpony) [reply tests/test_icecream.py:473]: Nope! Will remove. Good spot.

File: setup.py (hunk 9 above)
Comment (dawngerpony) [reply setup.py:106]: Noted.

File: icecream/icecream.py (hunk 8 above)
Comment (dawngerpony) [reply icecream/icecream.py:201]: Good point, will fix.

File: icecream/icecream.py (hunk 7 above)
Comment (dawngerpony) [reply icecream/icecream.py:173]: Will do on both counts.

File: icecream/icecream.py (hunk 6 above)
Comment (dawngerpony) [reply icecream/icecream.py:168]: Will do it that way then 👍

Comment (dawngerpony) [reply icecream/icecream.py:168]: (Although in my terminal, `env['COLUMNS']` isn't set, so I'm not sure which environments/OSes do set that variable)

Comment (alexmojaki) [reply icecream/icecream.py:168]: I don't know anything about this either, but I imagine it's a convention so that humans can easily set COLUMNS=50 when running a program and expect that many parts of the program will automatically respect that.

File: icecream/icecream.py [hunk 14]
```diff
@@ -34,6 +34,15 @@
...
 from .coloring import SolarizedDark
 
+try:
+    from shutil import get_terminal_size
+except ImportError:
+    try:
+        from backports.shutil_get_terminal_size import get_terminal_size
+    except ImportError:
+        def get_terminal_size():
+            return os.environ['COLUMNS']
```
Comment (alexmojaki) [inline icecream/icecream.py:44]:
```suggestion
        def get_terminal_size(fallback):
            columns = os.environ.get('COLUMNS', fallback[0])
            return (columns, 0)
```

File: icecream/icecream.py [hunk 15]
```diff
@@ -165,13 +174,13 @@ def detect_terminal_width(prefix, default=DEFAULT_LINE_WRAP_WIDTH):
...
     try:
-        if hasattr(shutil, "get_terminal_size"):
-            width = shutil.get_terminal_size().columns
-        else:  # Python 2.x doesn't support get_terminal_size
-            from backports.shutil_get_terminal_size import get_terminal_size
-            width = get_terminal_size().columns
-    except OSError:  # Not in TTY
+        # We need to pass a terminal height in the tuple so we pass the default
+        # of 25 lines but it's not used for anything.
+        width = get_terminal_size((default, 25)).columns
```
Comment (alexmojaki) [inline icecream/icecream.py:179]:
```suggestion
        width = get_terminal_size((default, 25))[0]
```

to work with the last resort `get_terminal_size` without having to involve namedtuples or something.

**Comment by alexmojaki** [comment]
Started to make some suggestions then I realised the `except Exception` is probably dealing with that more elegantly.

File: icecream/icecream.py [hunk 16]
```diff
@@ -154,15 +164,38 @@ def format_pair(prefix, arg, value):
...
 
+def detect_terminal_width(prefix, default=DEFAULT_LINE_WRAP_WIDTH):
+    """ Returns the number of columns that this terminal can handle. """
+    width = default
+    try:
+        # We need to pass a terminal height in the tuple so we pass the default
+        # of 25 lines but it's not used for anything.
+        width = get_terminal_size((default, 25)).columns
+    except Exception:  # Not in TTY or something else went wrong
+        pass
```
Comment (alexmojaki) [inline icecream/icecream.py:181]:
```suggestion
    try:
        # We need to pass a terminal height in the tuple so we pass the default
        # of 25 lines but it's not used for anything.
        width = get_terminal_size((default, 25)).columns
    except Exception:  # Not in TTY or something else went wrong
        width = default
```

**Comment by dawngerpony** [comment]
Hi @alexmojaki , sorry for the delay, work has been very busy. I have been thinking that it might be best not to enable this feature by default, at least in its first release, otherwise it could introduce breaking changes for a lot of people. So I have added some code to leave it off by default. There's now also at least one test that tests it by using a `@contextmanager` to set `env['COLUMNS']` to a specific value. Thoughts?

**Comment by alexmojaki** [comment]
> I have been thinking that it might be best not to enable this feature by default, at least in its first release, otherwise it could introduce breaking changes for a lot of people.

I think if the experience is generally improved then it should be enabled by default. Most people aren't going to even know they can configure this. As for breaking changes, this is a 'casual' debugging tool, I don't think there are significant expectations or guarantees of backward compatibility. It's hard to picture how this would significantly affect people negatively. If we want to be really careful, we could release a new major version. @gruns what do you think?

> NB. I'm also a little bemused by the current behaviour of my new `testLiteralWithShortTerminalWidth` test. I thought that when printing a literal, icecream wasn't supposed to include the "name of the variable" 
> Any idea what's going on?

This is a bug first noticed in https://github.com/gruns/icecream/issues/84#issuecomment-860132074

Basically the source of the literal argument is still printed when the value is multiline, but it shouldn't.

File: icecream/icecream.py [hunk 17]
```diff
@@ -193,6 +230,17 @@ def __call__(self, *args):
 
         return passthrough
 
+    def _setLineWrapWidth(self, detectTerminalWidth=False, terminalWidth=None):
+        prefix_length = len(self.prefix()) if callable(self.prefix) else len(self.prefix)
+        if terminalWidth:
+            width = terminalWidth
+        elif detectTerminalWidth is True:
+            width = detect_terminal_width(DEFAULT_LINE_WRAP_WIDTH)
```
Comment (alexmojaki) [inline icecream/icecream.py:238]:
```suggestion
            width = detect_terminal_width(DEFAULT_TERMINAL_WIDTH)
```

**Comment by alexmojaki** [comment]
The main problem is that you still need to account for the other parts of the output, and it's quite complicated. There's four pieces:

1. The global `prefix` (`ic| `)
2. The optional `context` (the filename, line number, etc)
3. The `argPrefix` (the source of the argument)
4. The value, which may not be included if the argument is a literal.

Each of these pieces needs to be added one at a time to see if they can fit within the terminal width. If not, that starts a new indented line with a new available width.

Your code mentions both 'terminal width' and 'line wrap width' in more places than I think it should, and the distinction is confusing. You should only need to store the attribute `lineWrapWidth`. It represents the full width for containing everything, by default the detected terminal width. The width passed to `argToStringFunction` should just be calculated from scratch each time `_constructArgumentOutput` is called. Don't calculate anything in advance based on `prefix`.

**Comment by dawngerpony** [comment]
> The main problem is that you still need to account for the other parts of the output, and it's quite complicated. There's four pieces:
> 
>     1. The global `prefix` (`ic| `)
> 
>     2. The optional `context` (the filename, line number, etc)
> 
>     3. The `argPrefix` (the source of the argument)
> 
>     4. The value, which may not be included if the argument is a literal.
> 
> 
> Each of these pieces needs to be added one at a time to see if they can fit within the terminal width. If not, that starts a new indented line with a new available width.

Thank you, this is the context I needed. I think I've catered for the prefix, but definitely not for anything else. I suspect that a lot more digging is needed for me to get familiar with how/when 2, 3 and 4 are added to the output.

> Your code mentions both 'terminal width' and 'line wrap width' in more places than I think it should, and the distinction is confusing. You should only need to store the attribute `lineWrapWidth`. It represents the full width for containing everything, by default the detected terminal width. The width passed to `argToStringFunction` should just be calculated from scratch each time `_constructArgumentOutput` is called. Don't calculate anything in advance based on `prefix`.

OK, will do.

Also, based on your thoughts, I'll refactor to remove the "disable by default". It'll be on for everyone. Let's just hope nobody is basing important logic on the output of an informal debugging tool 😉 

I'm not sure I picked the simplest task for my first contribution 😂  😬 

BTW, thank you so much for sticking with me on this 🙏

**Comment by Jakeroid** [comment]
@dawngerpony 

Hi,

Thank you for your efforts. Do you think you are still interested in contributing? I'm looking to clear PRs, Issues.
